### ISSUE TITLE
Client HttpRequest#path mishandles paths beginning with "//"

### ISSUE THREAD

**Issue by Zalathar** [eclipse/jetty.project#3098]
(Jetty version: 9.4.12.v20180830)

When a Jetty Client `Request` object has its path set by a call to `#path(String)`, and the specified path begins with `//`, the part of the path before the next `/` character is silently dropped.

Sample program:

```java
public static void main(String[] args) throws Exception {
    HttpClient client = new HttpClient();
    try {
        client.start();

        Request good = client.newRequest("http://example.test//first/second");
        System.out.printf("Good: <%s>\n", good.getPath());

        Request bad = client.newRequest("http://example.test");
        bad.path("//first/second"); // <-- issue occurs here
        System.out.printf("Bad:  <%s>\n", bad.getPath());
    } finally {
        client.stop();
    }
}
```

Expected output:
```
Good: <//first/second>
Bad:  <//first/second>
```

Actual output:
```
Good: <//first/second>
Bad:  </second>
```

(The same results occur when the request is actually sent over the network.)

**Comment by joakime** [comment]
It's technically difficult to handle `//`, as it has different meaning in different places.
The behavior you discovered in URI makes sense for the start of a path. (treating it as a scheme relative).

But lets say you have `/first//second/third` instead.
That's a path with 4 components (segments) of `"first", "", "second", "third"` and many libraries expect this behavior (as its dictated by both the HTTP and URI specs).

Some people view the existence of `//` as a mistake, similar to `/../` and want it "normalized" out of the URI.
But even [`URI.normalize()`](https://docs.oracle.com/javase/8/docs/api/java/net/URI.html#normalize--) does not remove `//` segments.

The syntax you encountered with `//<authority>/<path>` is standard URI spec for schemeless URIs.
The syntax rules you hit also applies to `<scheme>://<authority>/<path>`.

See https://tools.ietf.org/html/rfc3986#section-3

> When authority is present, the path must
> either be empty or begin with a slash ("/") character.  When
> authority is not present, the path cannot begin with two slash
> characters ("//").  

Which means even absolute URIs like you are using would be invalid if a scheme was attached.  
Eg: `http://first/second` would be interpreted by the `URI` parsing the same way.
a scheme of `http`, an authority of `first` and a path of `/second`.

In short, there's nothing "smart" that Jetty's HttpClient can do to interpret the "intent" of your URI and parse/strip/normalize what you see as a bad URI (it isn't bad to HttpClient, or the HTTP spec, or the URI spec).

**Comment by gregw** [comment]
Actually, I think jetty could be a little smarter here. 

 [RFC7230](https://tools.ietf.org/html/rfc7230#section-5.3) defines the target of a HTTP request as: 
```text
5.3.  Request Target

   Once an inbound connection is obtained, the client sends an HTTP
   request message (Section 3) with a request-target derived from the
   target URI.  There are four distinct formats for the request-target,
   depending on both the method being requested and whether the request
   is to a proxy.

     request-target = origin-form
                    / absolute-form
                    / authority-form
                    / asterisk-form
```

So a path of `//foo/bar` does not fit with `absolute-form`, `authority-form` or `asterisk-form`, thus is must be `origin-form` and the path should be set to literally `//foo/bar`..... ah but then we would run foul of https://tools.ietf.org/html/rfc3986#section-3

Hmmm so I think @joakime is correct and there is nothing smart we can do to satisfy both RFC!    I think  the URL with a path starting with // is just illegal one way or another.

**Comment by Zalathar** [comment]
> So a path of `//foo/bar` does not fit with `absolute-form`, `authority-form` or `asterisk-form`, thus is must be `origin-form` and the path should be set to literally `//foo/bar`..... ah but then we would run foul of https://tools.ietf.org/html/rfc3986#section-3

This is subtle, but I don't think that section's restriction is relevant here.

A URL without an authority part can't begin with `//`, not because it's forbidden, but because it's syntactically impossible. If you tried to write such a URL, you would end up with a scheme-relative URL instead.

However, an `origin-form` technically isn't a URL, and therefore doesn't have that syntactic restriction. It *looks* a lot like an authority-relative URL, and trying to parse it as one will often give the expected behaviour by accident, but in this corner-case it does not.

**Comment by gregw** [comment]
OK I agree that the path method could do a lot better job as it currently does not really handle most of those forms very well.
If the passed string starts with /, then we probably only need to split it on the ? looking for a query rather than build a full URI object. @sbordet thoughts?

However, we have to be careful to not break anything, so I'm not going to rush this into 9.4.14 and a fix will have to wait a few weeks for 9.4.15

**Comment by joakime** [comment]
The problem here seems to be on the `HttpClient` side, not the server side.

If you give the `HttpClient` this bad URI string of `//foo/bar`, it can't interpret "intent" either.
